File: scripts/init-project.mjs
#!/usr/bin/env node
import { access, mkdir, readFile, writeFile } from 'node:fs/promises';
import { constants } from 'node:fs';
import { dirname, resolve } from 'node:path';

const START = '<!-- workflow-os:v2:start -->';
const END = '<!-- workflow-os:v2:end -->';

function usage() {
  console.log('Usage: node scripts/init-project.mjs --target <project-path> [--name <name>] [--dry-run]');
}

function options(argv) {
  const values = new Map();
  for (let index = 0; index < argv.length; index += 1) {
    const token = argv[index];
    if (!token.startsWith('--')) continue;
    const [key, inline] = token.slice(2).split('=', 2);
    const next = argv[index + 1];
    values.set(key, inline ?? (next && !next.startsWith('--') ? (index += 1, next) : true));
  }
  return values;
}

async function exists(path) {
  try {
    await access(path, constants.F_OK);
    return true;
  } catch {
    return false;
  }
}

function managedGuidance() {
  return `${START}
# workflow-os v2

- 方向性 UI、数值、活动与技术选型先写入 \`docs/decisions/\`，负责人确认后才能实施。
- 调研、策划和实现可并行；同一文件或存在依赖的任务不可并行修改。
- 使用已安装的 \`wfos-core\` skill 路由能力；不可逆命令仍受 Codex/Claude 原生权限与项目 hook 约束。
- 运行 \`node <workflow-os-repo>/scripts/init-project.mjs\` 只补缺，不覆盖此项目的人类规则。
${END}
`;
}

async function copyMissing(path, content, dryRun, changes) {
  if (await exists(path)) {
    changes.push(`skip    ${path}`);
    return;
  }
  changes.push(`create  ${path}`);
  if (!dryRun) {
    await mkdir(dirname(path), { recursive: true });
    await writeFile(path, content, 'utf8');
  }
}

async function installGuidance(path, dryRun, changes) {
  const block = managedGuidance();
  if (!(await exists(path))) {
    changes.push(`create  ${path}`);
    if (!dryRun) await writeFile(path, block, 'utf8');
    return;
  }
  const current = await readFile(path, 'utf8');
  const start = current.indexOf(START);
  const end = current.indexOf(END);
  if ((start >= 0) !== (end >= 0) || (start >= 0 && end < start)) {
    throw new Error(`托管区标记损坏，拒绝修改：${path}`);
  }
  if (start >= 0) {
    const next = `${current.slice(0, start)}${block}${current.slice(end + END.length)}`;
    if (next !== current) changes.push(`update  ${path} (managed block)`);
    if (!dryRun && next !== current) await writeFile(path, next, 'utf8');
    return;
  }
  changes.push(`append  ${path} (managed block)`);
  if (!dryRun) await writeFile(path, `${current}${current.endsWith('\n') ? '\n' : '\n\n'}${block}`, 'utf8');
}

async function main() {
  const values = options(process.argv.slice(2));
  if (values.has('help') || !values.get('target')) {
    usage();
    process.exitCode = values.has('help') ? 0 : 1;
    return;
  }
  const root = resolve(String(values.get('target')));
  const name = String(values.get('name') ?? root.split(/[\\/]/).at(-1));
  const dryRun = values.has('dry-run');
  const changes = [];
  await copyMissing(`${root}/.workflow-os/project.yaml`, `schema: workflow-os/project@2\nname: ${JSON.stringify(name)}\ninitialized_at: ${JSON.stringify(new Date().toISOString())}\n`, dryRun, changes);
  await copyMissing(`${root}/docs/PROJECT.md`, `# ${name}\n\n## 目标\n\n- \n\n## 约束\n\n- \n`, dryRun, changes);
  await copyMissing(`${root}/docs/NOW.md`, '# 当前状态\n\n- \n', dryRun, changes);
  await copyMissing(`${root}/docs/INBOX.md`, '# 收集箱\n\n- \n', dryRun, changes);
  await copyMissing(`${root}/docs/decisions/README.md`, '# 决策\n\n方向性决策在确认前保持 pending。\n', dryRun, changes);
  await copyMissing(`${root}/docs/research/README.md`, '# 调研\n\n研究结论应保留来源与不确定性。\n', dryRun, changes);
  await installGuidance(`${root}/AGENTS.md`, dryRun, changes);
  console.log(`${dryRun ? 'Preview' : 'Initialized'} workflow-os project: ${root}`);
  for (const change of changes) console.log(change);
}

main().catch((error) => {
  console.error(`workflow-os init-project: ${error.message}`);
  process.exitCode = 1;
});

File: internal/initproject/initproject.go
// Package initproject 实现 `tessera init`:为项目补齐 Tessera 骨架文件,
// 只补缺、不覆盖已有内容。从 scripts/init-project.mjs 移植。
package initproject

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"
)

const (
	start = "<!-- tessera:v2:start -->"
	end   = "<!-- tessera:v2:end -->"
)

// Options 是 init 的解析后参数。
type Options struct {
	Target string
	Name   string
	DryRun bool
	Help   bool
	hasTgt bool
	hasNm  bool
}

// Parse 复刻 mjs 的参数解析:--k=v、--k v、--k(布尔)。
func Parse(args []string) Options {
	var o Options
	for i := 0; i < len(args); i++ {
		tok := args[i]
		if !strings.HasPrefix(tok, "--") {
			continue
		}
		key, inline, hasInline := strings.Cut(tok[2:], "=")
		val := ""
		isBool := true
		switch {
		case hasInline:
			val, isBool = inline, false
		case i+1 < len(args) && !strings.HasPrefix(args[i+1], "--"):
			val, isBool = args[i+1], false
			i++
		}
		switch key {
		case "target":
			o.Target, o.hasTgt = val, true
			_ = isBool
		case "name":
			o.Name, o.hasNm = val, true
		case "dry-run":
			o.DryRun = true
		case "help":
			o.Help = true
		}
	}
	return o
}

func managedGuidance() string {
	return start + `
# Tessera v2

- 方向性 UI、数值、活动与技术选型先写入 ` + "`docs/decisions/`" + `,负责人确认后才能实施。
- 调研、策划和实现可并行;同一文件或存在依赖的任务不可并行修改。
- 使用已安装的 ` + "`tessera-core`" + ` skill 路由能力;不可逆命令仍受 Codex/Claude 原生权限约束。
- 运行 ` + "`tessera init`" + ` 只补缺,不覆盖此项目的人类规则。
` + end + "\n"
}

func exists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}

func copyMissing(path, content string, dryRun bool, changes *[]string) error {
	if exists(path) {
		*changes = append(*changes, "skip    "+path)
		return nil
	}
	*changes = append(*changes, "create  "+path)
	if dryRun {
		return nil
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return err
	}
	return os.WriteFile(path, []byte(content), 0o644)
}

func installGuidance(path string, dryRun bool, changes *[]string) error {
	block := managedGuidance()
	if !exists(path) {
		*changes = append(*changes, "create  "+path)
		if !dryRun {
			return os.WriteFile(path, []byte(block), 0o644)
		}
		return nil
	}
	raw, err := os.ReadFile(path)
	if err != nil {
		return err
	}
	current := string(raw)
	s := strings.Index(current, start)
	e := strings.Index(current, end)
	if (s >= 0) != (e >= 0) || (s >= 0 && e < s) {
		return fmt.Errorf("托管区标记损坏,拒绝修改:%s", path)
	}
	if s >= 0 {
		next := current[:s] + block + current[e+len(end):]
		if next != current {
			*changes = append(*changes, "update  "+path+" (managed block)")
			if !dryRun {
				return os.WriteFile(path, []byte(next), 0o644)
			}
		}
		return nil
	}
	*changes = append(*changes, "append  "+path+" (managed block)")
	if !dryRun {
		sep := "\n\n"
		if strings.HasSuffix(current, "\n") {
			sep = "\n"
		}
		return os.WriteFile(path, []byte(current+sep+block), 0o644)
	}
	return nil
}

// Run 执行初始化,返回退出码。
func Run(args []string, out io.Writer) int {
	o := Parse(args)
	if o.Help || !o.hasTgt || o.Target == "" {
		fmt.Fprintln(out, "Usage: tessera init --target <project-path> [--name <name>] [--dry-run]")
		if o.Help {
			return 0
		}
		return 1
	}
	root, err := filepath.Abs(o.Target)
	if err != nil {
		fmt.Fprintln(out, "tessera init: "+err.Error())
		return 1
	}
	name := o.Name
	if !o.hasNm || name == "" {
		name = filepath.Base(root)
	}
	nameJSON, _ := json.Marshal(name)
	stamp := time.Now().UTC().Format("2006-01-02T15:04:05.000") + "Z"

	var changes []string
	files := []struct{ rel, content string }{
		{".tessera/project.yaml", fmt.Sprintf("schema: tessera/project@2\nname: %s\ninitialized_at: %q\n", nameJSON, stamp)},
		{"docs/PROJECT.md", fmt.Sprintf("# %s\n\n## 目标\n\n- \n\n## 约束\n\n- \n", name)},
		{"docs/NOW.md", "# 当前状态\n\n- \n"},
		{"docs/INBOX.md", "# 收集箱\n\n- \n"},
		{"docs/decisions/README.md", "# 决策\n\n方向性决策在确认前保持 pending。\n"},
		{"docs/research/README.md", "# 调研\n\n研究结论应保留来源与不确定性。\n"},
	}
	for _, f := range files {
		if err := copyMissing(filepath.Join(root, f.rel), f.content, o.DryRun, &changes); err != nil {
			fmt.Fprintln(out, "tessera init: "+err.Error())
			return 1
		}
	}
	if err := installGuidance(filepath.Join(root, "AGENTS.md"), o.DryRun, &changes); err != nil {
		fmt.Fprintln(out, "tessera init: "+err.Error())
		return 1
	}

	verb := "Initialized"
	if o.DryRun {
		verb = "Preview"
	}
	fmt.Fprintf(out, "%s Tessera project: %s\n", verb, root)
	for _, c := range changes {
		fmt.Fprintln(out, c)
	}
	return 0
}
